uaclient/daemon/retry_auto_attach.py

```diff
@@ -0,0 +1,183 @@
+    # in case we got started while already attached somehow
+    if cfg.is_attached:
+        return
+
+    # pick up where we left off
+    persisted_state = state_files.retry_auto_attach_state_file.read()
+    if persisted_state is not None:
+        # skip intervals we've already waited
+        offset = persisted_state.interval_index
+        intervals = RETRY_INTERVALS[offset:]
+        failure_reason = persisted_state.failure_reason
+    else:
+        offset = 0
+        intervals = RETRY_INTERVALS
+        failure_reason = None
+
+    for index, interval in enumerate(intervals):
+        last_attempt = datetime.datetime.now(datetime.timezone.utc)
+        next_attempt = last_attempt + datetime.timedelta(seconds=interval)
+        next_attempt = next_attempt.replace(second=0, microsecond=0)
+        state_files.retry_auto_attach_state_file.write(
+            state_files.RetryAutoAttachState(
+                interval_index=offset + index,
+                failure_reason=failure_reason,
+            )
+        )
+        msg_reason = failure_reason
+        if msg_reason is None:
+            msg_reason = "an unknown error"
+        try:
+            next_attempt = next_attempt.astimezone()
+        except Exception:
+            pass
+        auto_attach_status_msg = messages.AUTO_ATTACH_RETRY_NOTICE.format(
+            num_attempts=offset + index + 1,
+            reason=msg_reason,
+            next_run_datestring=next_attempt.isoformat(),
+        )
+        system.write_file(
+            AUTO_ATTACH_STATUS_MOTD_FILE, auto_attach_status_msg + "\n\n"
+        )
+        try:
+            with lock.SpinLock(
+                cfg=cfg,
+                lock_holder="pro.daemon.retry_auto_attach.notice_updates",
+            ):
+                cfg.notice_file.remove(
+                    "", messages.AUTO_ATTACH_RETRY_NOTICE_PREFIX
+                )
+                cfg.notice_file.add("", auto_attach_status_msg)
+        except exceptions.LockHeldError:
+            pass
+
+        time.sleep(interval)
+
+        if cfg.is_attached:
+            # We attached while sleeping - hooray!
+            break
+
+        try:
+            persisted_options = (
+                state_files.retry_auto_attach_options_file.read()
+            )
+            options = FullAutoAttachOptions()
+            if persisted_options is not None:
+                options.enable = persisted_options.enable
+                options.enable_beta = persisted_options.enable_beta
+            full_auto_attach(options)
+            break
+        except api_exceptions.AlreadyAttachedError:
+            LOG.info("already attached, ending retry service")
+            break
+        except api_exceptions.EntitlementsNotEnabledError as e:
+            LOG.warning(e.msg)
+            break
+        except Exception as e:
+            failure_reason = full_auto_attach_exception_to_failure_reason(e)
+            LOG.error(e)
+
+    cleanup(cfg)
```

Comment:
I could be missing something here, but if we cleanup the `retry_auto_attach_options_file` here, wouldn't we have a problem reading that file again during the retry ?

Reply:
This is after the retry loop so it should only happen after all retries have occurred, in which case we don't need the options or state file anymore.